### PR TITLE
Getting rid of nsenter and socat

### PR DESCRIPTION
**Pull request by abhi**
This commit removes the usage of nsenter and uses netns
to perform socat operation.
Fixes https://github.com/containerd/cri/issues/144

Signed-off-by: abhi <abhi@docker.com>

### CODE REVIEW

**Comment by Random-Liu** [comment]
@abhi Can we remove socat as well? I think we just need a goroutine to a tcp connection to the port and forward traffics.

**Comment by abhi** [comment]
@Random-Liu yes. Just raising the PR for e2e.

**Comment by Random-Liu** [comment]
@abhi I see. Will review it when it is ready. :)

File: pkg/server/sandbox_portforward.go [hunk 1]
```diff
@@ -46,69 +44,47 @@ func (c *criService) PortForward(ctx context.Context, r *runtime.PortForwardRequ
...
-	// Check following links for meaning of the options:
-	// * socat: https://linux.die.net/man/1/socat
-	// * nsenter: http://man7.org/linux/man-pages/man1/nsenter.1.html
-	args := []string{"-t", fmt.Sprintf("%d", pid), "-n", socat,
-		"-", fmt.Sprintf("TCP4:localhost:%d", port)}
-
-	nsenter, err := exec.LookPath("nsenter")
-	if err != nil {
-		return errors.Wrap(err, "failed to find nsenter")
+	if s.NetNS == nil {
```
Comment (Random-Liu) [inline pkg/server/sandbox_portforward.go]: || s.NetNS.Closed()

File: pkg/server/sandbox_portforward.go [hunk 2]
```diff
@@ -46,69 +44,47 @@ func (c *criService) PortForward(ctx context.Context, r *runtime.PortForwardRequ
...
+		}()
+		wg.Add(1)
+		go func() {
+			if _, err := io.Copy(stream, client); err != nil {
+				logrus.WithError(err).Errorf("Failed to copy port forward output for %q port %d", id, port)
+			}
+			wg.Done()
+		}()
+		wg.Wait()
+		logrus.Infof("Finish copy port forward input for %q port %d: %v", id, port)
```
Comment (Random-Liu) [inline pkg/server/sandbox_portforward.go]: Add a log for both input copy and output copy, and move them into goroutine?

File: pkg/server/sandbox_portforward.go [hunk 3]
```diff
@@ -46,69 +44,47 @@ func (c *criService) PortForward(ctx context.Context, r *runtime.PortForwardRequ
...
-
-	if err := cmd.Run(); err != nil {
-		return errors.Errorf("nsenter command returns error: %v, stderr: %q", err, stderr.String())
+		defer client.Close()
+
+		wg.Add(1)
+		go func() {
+			if _, err := io.Copy(client, stream); err != nil {
+				logrus.WithError(err).Errorf("Failed to copy port forward input from %q port %d", id, port)
+			}
```
Comment (Random-Liu) [inline pkg/server/sandbox_portforward.go]: client.Close()?

File: pkg/server/sandbox_portforward.go [hunk 4]
```diff
@@ -46,69 +44,47 @@ func (c *criService) PortForward(ctx context.Context, r *runtime.PortForwardRequ
...
+			wg.Done()
+		}()
+		wg.Add(1)
+		go func() {
+			if _, err := io.Copy(stream, client); err != nil {
+				logrus.WithError(err).Errorf("Failed to copy port forward output for %q port %d", id, port)
+			}
+			wg.Done()
+		}()
```
Comment (Random-Liu) [inline pkg/server/sandbox_portforward.go]: stream.Close()?

Are these 2 `Close()` are necessary?
If `dst` is closed, `src` is open, but there is no new input, will `io.Copy` stop?
* If it does, we don't need the 2 `Close()`.
* If it doesn't, I think we'll probably need them.

File: pkg/server/sandbox_portforward.go [hunk 5]
```diff
@@ -17,24 +17,22 @@ limitations under the License.
 package server
```
Comment (Random-Liu) [inline pkg/server/sandbox_portforward.go]: Update dependencies in `README.md`.

File: pkg/server/sandbox_portforward.go (hunk 1 above)
Comment (abhi) [reply pkg/server/sandbox_portforward.go]: thats already checked by netns.Do

File: pkg/server/sandbox_portforward.go (hunk 3 above)
Comment (abhi) [reply pkg/server/sandbox_portforward.go]: I have added a defer on top

File: pkg/server/sandbox_portforward.go (hunk 2 above)
Comment (abhi) [reply pkg/server/sandbox_portforward.go]: sure

**Comment by abhi** [comment]
@Random-Liu done.

**Comment by Random-Liu** [comment]
@abhi We need to start 2 goroutines to copy stream. However, the 2 goroutines may not be in the same network namespace. We are not sure whether it will work or not. The test result shows that it works, but we need to double check with the golang team.

I don't think we have time to figure that out in the coming `rc.0` release. @abhi Let's only remove `nsenter` and still use `socat` in this PR. We can remove `socat` in `rc.1` release.

**Comment by abhi** [comment]
@Random-Liu yes I am of the same opinion as well. I will remove the socat commits.

**Comment by Random-Liu** [comment]
@abhi also pointed me to another potential issue https://github.com/containernetworking/plugins/issues/98.

Given so, let's move this issue to next release, so that we have more time to investigate this.

It is safe to move this PR to next release, because there is no actual functionality change, we just get rid of 2 binary dependencies `nsenter` and `socat` in this PR.

**Comment by Random-Liu** [comment]
We plan to cut an early `rc.1` release today to carry several fix and improvement. Move this to `rc.2`.

**Comment by Random-Liu** [comment]
@abhi Here is a piece of code:
```golang
package main

import (
	"fmt"
	"io"
	"os"
	"sync"
	"time"
)

func main() {
	f, err := os.OpenFile("io", os.O_RDWR|os.O_CREATE, 0666)
	if err != nil {
		panic(err)
	}
	var wg sync.WaitGroup
	wg.Add(2)
	go func() {
		io.Copy(os.Stdout, f)
		fmt.Println("file -> stdout done")
		wg.Done()
	}()
	go func() {
		io.Copy(f, os.Stdin)
		fmt.Println("stdin -> file done")
		wg.Done()
	}()
	time.Sleep(5 * time.Second)
	f.Close()
	wg.Wait()
	fmt.Println("all done")
}
```

You will see, after 5 seconds, the problem won't exit even though `f` is closed. Because there is no input from `os.Stdin` to drive the `io.Copy` to write once, and `io.Copy` won't know `f` has been closed. Once you input something into stdin, the problem will exit.

This is why I said, you need to close the other side when one `io.Copy` is done.

File: pkg/server/sandbox_portforward.go (hunk 1 above)
Comment (Random-Liu) [reply pkg/server/sandbox_portforward.go]: Our `Closed` is different https://github.com/containerd/cri/blob/master/pkg/store/sandbox/netns.go#L108.

We have the `restore` support.

File: pkg/server/sandbox_portforward.go [hunk 6]
```diff
@@ -46,69 +44,48 @@ func (c *criService) PortForward(ctx context.Context, r *runtime.PortForwardRequ
...
-	// * socat: https://linux.die.net/man/1/socat
-	// * nsenter: http://man7.org/linux/man-pages/man1/nsenter.1.html
-	args := []string{"-t", fmt.Sprintf("%d", pid), "-n", socat,
-		"-", fmt.Sprintf("TCP4:localhost:%d", port)}
-
-	nsenter, err := exec.LookPath("nsenter")
-	if err != nil {
-		return errors.Wrap(err, "failed to find nsenter")
+	if s.NetNS == nil {
+		return errors.Errorf("failed to find network namespace fo sandbox %q in store", id)
```
Comment (Random-Liu) [inline pkg/server/sandbox_portforward.go]: network namespace for sandbox %q is closed

File: pkg/server/sandbox_portforward.go [hunk 7]
```diff
@@ -46,69 +44,48 @@ func (c *criService) PortForward(ctx context.Context, r *runtime.PortForwardRequ
...
-		return errors.Wrap(err, "failed to create stdin pipe")
-	}
-	go func() {
-		if _, err := io.Copy(in, stream); err != nil {
-			logrus.WithError(err).Errorf("Failed to copy port forward input for %q port %d", id, port)
+	err = s.NetNS.GetNs().Do(func(_ ns.NetNS) error {
+		var wg sync.WaitGroup
+		client, err := net.Dial("tcp4", fmt.Sprintf("localhost:%d", port))
+		if err != nil {
+			return errors.Wrap(err, "failed to dial")
```
Comment (Random-Liu) [inline pkg/server/sandbox_portforward.go]: nit: "failed to dial %q", port

File: pkg/server/sandbox_portforward.go [hunk 8]
```diff
@@ -46,69 +44,48 @@ func (c *criService) PortForward(ctx context.Context, r *runtime.PortForwardRequ
...
-		logrus.Debugf("Finish copy port forward input for %q port %d", id, port)
-	}()
-
-	if err := cmd.Run(); err != nil {
-		return errors.Errorf("nsenter command returns error: %v, stderr: %q", err, stderr.String())
+		wg.Add(1)
+		go func() {
+			defer client.Close()
+			if _, err := io.Copy(client, stream); err != nil {
+				logrus.WithError(err).Errorf("Failed to copy port forward input from %q port %d", id, port)
```
Comment (Random-Liu) [inline pkg/server/sandbox_portforward.go]: s/from/for?

File: pkg/server/sandbox_portforward.go [hunk 9]
```diff
@@ -46,69 +44,48 @@ func (c *criService) PortForward(ctx context.Context, r *runtime.PortForwardRequ
...
+			}
+			logrus.Infof("Finish copy port forward input for %q port %d: %v", id, port)
```
Comment (Random-Liu) [inline pkg/server/sandbox_portforward.go]: remove %v

File: pkg/server/sandbox_portforward.go [hunk 10]
```diff
@@ -46,69 +44,48 @@ func (c *criService) PortForward(ctx context.Context, r *runtime.PortForwardRequ
...
+			wg.Done()
+		}()
+		wg.Add(1)
+		go func() {
+			defer stream.Close()
+			if _, err := io.Copy(stream, client); err != nil {
+				logrus.WithError(err).Errorf("Failed to copy port forward output for %q port %d", id, port)
+			}
+			logrus.Infof("Finish copy port forward output for %q port %d: %v", id, port)
```
Comment (Random-Liu) [inline pkg/server/sandbox_portforward.go]: ditto.

**Pull request merged**